File: analytics/src/main/java/io/appmetrica/analytics/impl/AppStandbyBucketConverter.java
package io.appmetrica.analytics.impl;

import android.annotation.TargetApi;
import android.app.usage.UsageStatsManager;
import android.os.Build;
import androidx.annotation.Nullable;
import io.appmetrica.analytics.coreutils.internal.AndroidUtils;

public class AppStandbyBucketConverter {

    @Nullable
    public BackgroundRestrictionsState.AppStandByBucket fromIntToAppStandbyBucket(final int value) {
        if (AndroidUtils.isApiAchieved(Build.VERSION_CODES.P)) {
            return fromIntToStandbyBucketP(value);
        } else {
            return null;
        }
    }

    @TargetApi(Build.VERSION_CODES.P)
    @Nullable
    private BackgroundRestrictionsState.AppStandByBucket fromIntToStandbyBucketP(final int value) {
        if (AndroidUtils.isApiAchieved(Build.VERSION_CODES.R)) {
            if (value == UsageStatsManager.STANDBY_BUCKET_RESTRICTED) {
                return BackgroundRestrictionsState.AppStandByBucket.RESTRICTED;
            }
        }
        switch (value) {
            case UsageStatsManager.STANDBY_BUCKET_ACTIVE:
                return BackgroundRestrictionsState.AppStandByBucket.ACTIVE;
            case UsageStatsManager.STANDBY_BUCKET_FREQUENT:
                return BackgroundRestrictionsState.AppStandByBucket.FREQUENT;
            case UsageStatsManager.STANDBY_BUCKET_WORKING_SET:
                return BackgroundRestrictionsState.AppStandByBucket.WORKING_SET;
            case UsageStatsManager.STANDBY_BUCKET_RARE:
                return BackgroundRestrictionsState.AppStandByBucket.RARE;
            default:
                return null;
        }
    }

    @Nullable
    public String fromAppStandbyBucketToString(@Nullable BackgroundRestrictionsState.AppStandByBucket standByBucket) {
        if (standByBucket == null) {
            return null;
        }
        switch (standByBucket) {
            case ACTIVE:
                return "ACTIVE";
            case WORKING_SET:
                return "WORKING_SET";
            case FREQUENT:
                return "FREQUENT";
            case RARE:
                return "RARE";
            case RESTRICTED:
                return "RESTRICTED";
            default:
                return null;
        }
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/AppStandbyBucketConverter.kt
package io.appmetrica.analytics.impl

import android.annotation.TargetApi
import android.app.usage.UsageStatsManager
import android.os.Build
import io.appmetrica.analytics.coreutils.internal.AndroidUtils.isApiAchieved
import io.appmetrica.analytics.impl.BackgroundRestrictionsState.AppStandByBucket
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger

internal class AppStandbyBucketConverter {

    private val tag = "[AppStandbyBucketConverter]"

    private val usageStateManagerAppStandByBucketExempted = 5

    fun fromIntToAppStandbyBucket(value: Int): AppStandByBucket? {
        return if (isApiAchieved(Build.VERSION_CODES.P)) {
            value.fromIntToStandbyBucketP()
        } else {
            null
        }
    }

    @TargetApi(Build.VERSION_CODES.P)
    private fun Int.fromIntToStandbyBucketP(): AppStandByBucket? {
        DebugLogger.info(tag, "fromIntToStandbyBucketP: %s", this)

        if (isApiAchieved(Build.VERSION_CODES.R)) {
            if (this == UsageStatsManager.STANDBY_BUCKET_RESTRICTED) {
                return AppStandByBucket.RESTRICTED
            }
        }
        return when (this) {
            usageStateManagerAppStandByBucketExempted -> AppStandByBucket.EXEMPTED
            UsageStatsManager.STANDBY_BUCKET_ACTIVE -> AppStandByBucket.ACTIVE
            UsageStatsManager.STANDBY_BUCKET_FREQUENT -> AppStandByBucket.FREQUENT
            UsageStatsManager.STANDBY_BUCKET_WORKING_SET -> AppStandByBucket.WORKING_SET
            UsageStatsManager.STANDBY_BUCKET_RARE -> AppStandByBucket.RARE
            else -> null
        }
    }

    fun fromAppStandbyBucketToString(standByBucket: AppStandByBucket?): String? {
        if (standByBucket == null) {
            return null
        }
        return when (standByBucket) {
            AppStandByBucket.EXEMPTED -> "EXEMPTED"
            AppStandByBucket.ACTIVE -> "ACTIVE"
            AppStandByBucket.WORKING_SET -> "WORKING_SET"
            AppStandByBucket.FREQUENT -> "FREQUENT"
            AppStandByBucket.RARE -> "RARE"
            AppStandByBucket.RESTRICTED -> "RESTRICTED"
            else -> null
        }
    }
}
